### ISSUE TITLE
i2c read returns data from "read-1"

### ISSUE THREAD

**Issue by akohlsmith** [eriksl/usitwislave#6]
I'm sure I'm doing something stupid, but I can't figure it out.

I've implemented a typical "register based" I2C slave.  It follows the standard practise of a write followed by a restart and read. In BusPirate parlance:

`[ 0x10 0x05 { 0x11 rr ]`

(start, address 0x10, register 5, restart, address 0x11 read 2 bytes stop)

My code sets a "register number" variable if the callback has any data coming in, and then always looks at this register number to determine what to fill the output buffer with.

What I am seeing is that if I want to read from register 1 and then register 2, the read for register 2 will return register 1's data the first time. The second read from register 2 will return register 2's data. It's like the response back to the master lags by one transaction.

e.g. if register 1 holds the value 0x1234 and register 2 holds the value 0x5678, then again in BusPirate parlance:

`[ 0x10 0x01 { 0x11 rr ]` returns register 0's contents (assuming reset/power on)

`[ 0x10 0x02 { 0x11 rr ]` returns 0x12 0x34 (register 1's contents)

`[ 0x10 0x02 { 0x11 rr ]` returns 0x56 0x78 (register 2's contents)

i.e. the returned data seems to "lag" by one transaction. I return the previous write's data.

My code seems that it should work:

```c
static void twi_callback(uint8_t inlen, const uint8_t *in, uint8_t *outlen, uint8_t *out)
{
	static uint8_t reg;

	/* do we have at least one byte to parse? */
	if (inlen) {
		switch (in[0]) {
		case 0x01:
		case 0x02:
			reg = in[0];
			break;

		default:
			/* invalid register selected, do nothing */
			break;
		};
	}

	/*
	 * determine what to send out
	 * 'reg' has the register number that was requested.
	 * it's guaranteed to be a valid register.
	 */
	switch (reg) {
	case 0x01:
		out[0] = 0x12;
		out[1] = 0x34;
		*outlen = 2;
		break;

	case 0x02:
		out[0] = 0x56;
		out[1] = 0x78;
		*outlen = 2;
		break;

	/* default to indicate something odd, no data returned */
	default:
		out[0] = 0xaa;
		out[1] = 0xaa;
		*outlen = 0;
		break;
	}
}
```

As you can see, out and outlen are always initialized and they should be "current".

**Comment by eriksl** [comment]
Yes, reSTARTS are not supported, for a large part due to the limitations of USI. So make it two separate transactions: START set register STOP START read register STOP.

**Comment by akohlsmith** [comment]
Hm, I don't think that is the reason.

I set up a register to update its count every second or so.  I issue reads (just a simple read, no restarts in sight) and it works... but if I leave it for a period of time and execute a read, it returns the count that would have been before I left it alone. A second read "jumps" ahead to where I expect it to be.

i.e.

`[ 0x11 rr ]` returns 0x15 0x00
`[ 0x11 rr ]` returns 0x17 0x00
`[ 0x11 rr ]` returns 0x1a 0x00

Now leave it alone for a minute and execute two more reads:

`[ 0x11 rr ]` returns 0x1c 0x00
`[ 0x11 rr ]` returns 0x30 0x00

It's the same kind of problem (I get the "last read"'s data).

**Comment by eriksl** [comment]
The problem could be that after the first sequence the tiny needs to do some processing and isn't ready in time to respond to the second sequence with fresh data?

I've had this working for years (and still runs), send i2c command and read back the reply, so it can't be a major problem ;-) My I2C master can't issue i2c sequences at full bus speed (USB to I2C converter) that may be an explanation.

**Comment by penguineer** [comment]
I have the same issue: All works fine on a Raspberry Pi A, but the same device fails on a Raspberry Pi 2B – Response to the 2nd query is actually the response to the 1st query. I'm doing this manually and it works even if there are several minutes delay in between.

**Comment by penguineer** [comment]
I think the problem is described here: https://community.atmel.com/forum/e70-twi-master-read-communication-problem

**Comment by eriksl** [comment]
There is a principal problem with USI (for both SPI and I2C). As a master it's okay, but as a slave there are very tight timing demands, because the slave doesn't control the clock and must respond within each clock cycle for some events. For simple data output this is performed by the shift register, but for things like address recognition, it must be done very quickly. Also if an ACK is done too late, it will be interpreted by the master as a NAK. For a bus running at the recommend standard speed of 100 kHz, an ATmel USI client running at 8 or even 1 MHz is really too slow to always react in time, is my experience.

Then there is the master implementation on the other side. Ideally the master should assert SDA just in the middle of an SCL low-to-high-to-low transition and it should sample SDA just in the middle of a high-to-low-to-high transition. A hardware implementation will do that by running a clock 4 times (or more) the SCL clock frequency. Software (bit banging) I2C implementations are often sloppy at this point. I often see SDA being asserted and sampled just after/before SCL transition. This will make the whole thing less robust.

So my advice is to either stick to master mode (which is ok) or use a device that has a full I2C implemenation, like ATmega, I've never had issues there. The ATmega's I2C is really quite limited still, it can only send one byte at a time. Some newer microcontrollers can have most of the slave functionality in hardware, where you only have to fill certain memory areas with the data.
